runtime/parachains/src/origin.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Declaration of the parachain specific origin and a pallet that hosts it.
+
+use sp_std::result;
+use sp_runtime::traits::BadOrigin;
+use primitives::v1::Id as ParaId;
+use codec::{Decode, Encode};
+
+/// Origin for the parachains.
+#[derive(PartialEq, Eq, Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Origin {
+	/// It comes from a parachain.
+	Parachain(ParaId),
+}
+
+/// Ensure that the origin `o` represents a parachain.
+/// Returns `Ok` with the parachain ID that effected the extrinsic or an `Err` otherwise.
+pub fn ensure_parachain<OuterOrigin>(o: OuterOrigin) -> result::Result<ParaId, BadOrigin>
+	where OuterOrigin: Into<result::Result<Origin, OuterOrigin>>
+{
+	match o.into() {
+		Ok(Origin::Parachain(id)) => Ok(id),
+		_ => Err(BadOrigin),
+	}
+}
+
+/// The origin module.
+pub trait Trait: frame_system::Trait {}
+
+frame_support::decl_module! {
+	/// There is no way to register an origin type in `construct_runtime` without a pallet the origin
+	/// belongs to.
+	///
+	/// This module fulfills only the single purpose of housing the `Origin` in `construct_runtime`.
```

Comment:
Sounds like you want to open an issue in Substrate :D

Reply:
Thanks for the suggestion, I left an issue https://github.com/paritytech/substrate/issues/7203